src/bayesdb.py

```diff
@@ -377,6 +377,19 @@ def changes(self):
         """
         return self._sqlite3.changes()
 
+    def dump_models(self, population_name, path):
```

Comment:
This function should not be a member of the `bayesdb.py` class, which is a lightweight class on top of a sqlite3 handle connection and not designed to handle ad-hoc operations on generators, populations, etc., and certainly not writing to disk.

Instead the user should write custom model dumping code in their own modules and doing their own checks (such as the assertion that only a population with one generator can be dumped to JSON) and their own formatting conventions (e.g., using indentation=2) by using the bayesdb API, e.g., as follows
```python
def dump_models(bdb, population_name, path):
    population_id = core.bayesdb_get_population(bdb, population_name)
    generators = core.bayesdb_population_generators( bdb, population_id)
    assert len(generators) == 1
    generator_id = generators[0]
    backend = core.bayesdb_generator_backend(bdb, generator_id)
    j = backend.json_ready_models(bdb, generator_id)
    with open(path, 'w') as outfile:
        json.dump(j, outfile, indent=2)

```

I further suggest that the client changes `path` to `stream` so that the JSON can be written to any object on which `json.dump` knows how to operate (e.g., `StringIO` or others).

Reply:
@fsaad RE `Instead the user should write custom model dumping code`: While I agree with your objection in principle, I don't see a way to align the objection with the lab's plans for the workshops, the class, and the center partners and thus I don't see a way around adding this functionality to the Bayeslite code base and enforce assertions based on restrictions regarding the number of generators. We can't expect a user who is not a sincere contributor to the code base to learn how to use and write the code in the snippet you have suggested. We further don't have the capacity to add any additional packages to Conda because of the overhead it produces. The combination of both cause me to believe that there is no way around adding this to Bayeslite.

Reply:
Which software is the expected client of `def dump_models`?

For example, if the client is iventure, then a model dumping snippet similar to the one above should be implemented there and likely exposed as a magic `%dump_models`.

If the client is `crosscat_to_metaprob`, then a model dumping procedure should be added there, suited to the specific use case.

If the client is writing BQL/MML code, then the model dumping procedure should be exposed in a manner consistent with other operations in the interpreter from `bql.py`, e.g., `EXPORT <population_name>|<generator_name> TO <path>`.

If the client is an arbitrary user who is interacting with bayeslite in a python script or a python library (which I don't think is the case for workshop, class, or partners; but will be the case for e.g., a webserver or other layers on top of bayeslite), then it is the expectation that they know how to dump models, since they are already leveraging the bayeslite API.

Whatever the case, polluting the `BayesDB` class in `bayesdb.py`, which provides a lightweight wrapper and generic wrapper on top of a sqlite connection, with an ad-hoc model dumping routine is not the correct approach.  At the very least, any implementation (which still isn't clear to me that bayeslite should provide an ad-hoc function for writing code to disk, the backend-specific `json_ready_models` already provides the necessary information) should be a in specific module dedicated to that purpose, e.g., `src/export.py` or similar.

Reply:
Migrating [this bit](https://github.com/jar398/bayeslite/blob/20191202-rees-dump-models/src/bayesdb.py#L380-L391) to [iVenture](https://github.com/probcomp/iventure) is fine with me.

Reply:
Keeping the model dumping code out of the bayeslite repo is perfectly fine with me - I have no skin in the game. In that case this PR should simply be rejected in its entirety. But I will need precise advice as to where the code properly belongs. As far as I can tell the two of you have not resolved this question between you.

Reply:
@jar398 Ulrich and I have reached agreement---this code snipped that writes to disk will be moved to iventure.